src/main/java/de/retest/recheck/configuration/ProjectConfiguration.java

```diff
@@ -0,0 +1,88 @@
+package de.retest.recheck.configuration;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProjectConfiguration {
+
+	private static final Logger logger = LoggerFactory.getLogger( ProjectConfiguration.class );
+	private static final String RETEST_PROJECT_DEFAULTS = "retest-project-defaults.properties";
+	private static ProjectConfiguration INSTANCE;
+
+	public static final String RETEST_PROJECT_PROPERTIES = "retest.properties";
+	public static final String RETEST_PROJECT_CONFIG_FOLDER = ".retest";
+
+	private ProjectConfiguration() {
+
+	}
+
+	public static ProjectConfiguration getInstance() {
+		if ( INSTANCE == null ) {
+			INSTANCE = new ProjectConfiguration();
+		}
+		return INSTANCE;
+	}
+
+	public void initializeConfiguration() {
+		final Path configFolder = Paths.get( "" ).toAbsolutePath().resolve( RETEST_PROJECT_CONFIG_FOLDER );
+		createProjectConfigurationFolderIfNeeded( configFolder );
+		createEmptyProjectConfigurationIfNeeded( configFolder );
+	}
+
+	private void createProjectConfigurationFolderIfNeeded( final Path configFolder ) {
+		if ( !configFolder.toFile().exists() ) {
+			try {
+				Files.createDirectories( configFolder );
+				Files.copy( getClass().getClassLoader().getResourceAsStream( RETEST_PROJECT_DEFAULTS ),
+						configFolder.resolve( RETEST_PROJECT_PROPERTIES ) );
+				logger.info( "Creating empty project configuration in {}.", configFolder );
+			} catch ( final IOException e ) {
+				logger.error( "Error creating project configuration folder in {}.", configFolder );
+			}
+		}
+	}
+
+	private void createEmptyProjectConfigurationIfNeeded( final Path configFolder ) {
+		final Path configFile = configFolder.resolve( RETEST_PROJECT_PROPERTIES );
+		if ( !configFile.toFile().exists() ) {
+			try {
+				Files.copy( getClass().getClassLoader().getResourceAsStream( RETEST_PROJECT_DEFAULTS ), configFile );
+				logger.info( "Creating empty recheck configuration in {}.", configFile );
+			} catch ( final IOException e ) {
+				logger.error( "Error creating empty recheck configuration in {}.", configFile );
+			}
+		}
+	}
+
+	public static Path getProjectRoot( final Path basePath ) {
+		Path currentFolder = basePath;
+
+		if ( !currentFolder.toFile().isDirectory() ) {
+			currentFolder = currentFolder.getParent();
+		}
+
+		while ( !containsRetestConfig( currentFolder ) ) {
+			currentFolder = currentFolder.getParent();
+		}
+
+		return currentFolder;
+	}
+
+	private static boolean containsRetestConfig( final Path basePath ) {
+		try ( Stream<Path> filesInBasePath = Files.list( basePath ) ) {
```

Comment:
Is this a resource that needs to be closed? If not, how about:

```java
try {
    return Files.list( basePath ) //
            .map( Path::getFileName ) //
            .filter( Objects::nonNull ) //
            .map( Path::toString ) //
            .anyMatch( p -> p.equals( RETEST_PROJECT_CONFIG_FOLDER ) );
} catch ( ... ) { ...
```

Reply:
That's what sonartype told me :)

Reply:
`Files.list` automatically registers a stream close listener that closes the file.